File: python-algorithms/src/add_two_numbers.py
from typing import Optional

from src.ListNode import ListNode

class AddTwoNumbers:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = None
        tail = None
        carry = False

        while not (l1 is None and l2 is None):
            a = 0
            b = 0

            if l1 is not None:
                a = l1.val
                l1 = l1.next
            if l2 is not None:
                b = l2.val
                l2 = l2.next

            sum = a + b
            
            if carry:
                sum += 1
                carry = False

            if sum > 9:
                carry = True
                sum -= 10

            new_node = ListNode(sum)

            if head is None:
                head = new_node
                tail = new_node
            else:
                tail.next = new_node
                tail = tail.next


        if carry:
            tail.next = ListNode(1)

        return head

File: java-algorithms/src/main/java/com/omsprog/algorithms/lc/AddTwoNumbers.java
package com.omsprog.algorithms.lc;

public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = null;
        ListNode tail = null;
        boolean carry = false;

        while(!(l1 == null && l2 == null)) {
            int a = 0;
            int b = 0;
            int sum = 0;

            if(l1 != null) {
                a = l1.val;
                l1 = l1.next;
            }

            if(l2 != null) {
                b = l2.val;
                l2 = l2.next;
            }

            sum = a + b;

            if(carry) {
                ++sum;
                carry = false;
            }

            if(sum > 9) {
                sum = sum - 10;
                carry = true;
            }

            ListNode newNode = new ListNode(sum);

            if(head == null) {
                head = newNode;
                tail = newNode;
            } else {
                tail.next = newNode;
                tail = tail.next;
            }
        }

        if(carry)
            tail.next = new ListNode(1);

        return head;
    }
}
